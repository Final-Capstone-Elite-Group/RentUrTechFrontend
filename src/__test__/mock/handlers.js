import { rest } from 'msw';

const handlers = [
  rest.get('http://localhost:3000/equipments', (req, res, ctx) => res(
    ctx.json({
      data: [
        {
          id: '14',
          type: 'equipment',
          attributes: {
            id: 14,
            title: 'Wireless Printer',
            description: "The Wireless Printer gives you all the essential features you need—easily print, scan and copy, and activate HP+: HP's smart printing system that comes with 6 months of free Instant Ink and a 2 year extended HP warranty.",
            review: 'https://www.youtube.com/watch?v=9Xk39LNg304',
            duration: 5,
            rent_fee: '821.0',
            dates_reserved: [],
            total_amount_payable: '22.0',
            created_at: '2022-06-08T19:42:08.103Z',
            updated_at: '2022-06-08T19:42:08.112Z',
            image: {
              url: 'http://localhost:3000/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBJUT09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--2c533787f3499541cdd2bd6da005bace4749ddb7/wireless_printer.jpeg',
            },
          },
        },
        {
          id: '13',
          type: 'equipment',
          attributes: {
            id: 13,
            title: 'Turntable',
            description: 'Whether you’re just getting into records, rediscovering a record collection long since buried away or simply adding a turntable to complete your system, the AT-LP60X fully automatic belt-drive turntable is an excellent choice. This affordable, easy-to-use turntable plays both 33-1/3 and 45 RPM records and, with its built-in switchable phonon preamp, can be connected directly to your home stereo, powered speakers, computer and other components, whether they have a dedicated phonon input or not. An update of the popular AT-LP60, the ATLP60X features a redesigned tone arm base and head shell to improve tracking and reduce resonance. Noise is further reduced by the inclusion of an AC adapter that moves the AC/DC conVersion outside of the chassis where its impact on the signal chain is limited. Audio-Technical has been a leader in phonon cartridge design for more than 50 years, and that expertise shows in the AT-LP60X’s Dual Magnet cartridge with replaceable stylus. The cartridge is integrated into the head shell for assured performance and ease of setup. The turntable comes with a detachable RCA output cable (3. 5 mm male to dual RCA male), 45 RPM adapter, and a removable hinged dust cover. It is available in black (AT-LP60X-BK), brown/black (AT-LP60X-BW), gunmetal/black (AT-LP60X-GM), and red/black (AT-LP60X-RD).',
            review: 'https://www.youtube.com/watch?v=Yaj1Mhj8mo4',
            duration: 6,
            rent_fee: '832.0',
            dates_reserved: [],
            total_amount_payable: '26.0',
            created_at: '2022-06-08T19:42:08.067Z',
            updated_at: '2022-06-08T19:42:08.081Z',
            image: {
              url: 'http://localhost:3000/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBJQT09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--ee3004a8fe9fce926a806f65e4a87b9aa25c45af/turntable.jpeg',
            },
          },
        }],
    }),
    ctx.status(200),
  )),

  rest.get('http://localhost:3000/reservations', (req, res, ctx) => res(
    ctx.json({
      data: [
        {
          id: '15',
          type: 'reservation',
          attributes: {
            id: 15,
            total: '52.0',
            reserved_date: '2022-07-26',
            city: 'Somewhere nice',
            created_at: '2022-06-09T15:07:15.493Z',
            updated_at: '2022-06-09T15:07:15.493Z',
            equipment: {
              id: 4,
              title: 'GPS Drone',
              url: 'http://localhost:3000/rails/active_storage/blobs/redirect/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBGdz09IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--9d122d8fb7af2f04562aa6a26d728ed7d3fe8c64/drone1.jpg',
              duration: 5,
            },
          },
        }],
    }),
    ctx.status(200),
  )),

  rest.post('http://localhost:3000/reservations', (req, res, ctx) => res(
    ctx.json(
      {
        id: 12,
        total: '52.0',
        reserved_date: '2022-06-09T00:00:00.000Z',
        city: 'Somewhere nice',
        equipment_id: 4,
        user_id: 25,
        created_at: '2022-06-09T15:02:25.381Z',
        updated_at: '2022-06-09T15:02:25.381Z',
      },
    ),
    ctx.status(200),
  )),

  rest.delete('http://localhost:3000/reservations/15', (req, res, ctx) => res(
    ctx.json(
      {
        data: 'Reservation destroyed successfully',
      },
    ),
    ctx.status(200),
  )),
];

export default handlers;
