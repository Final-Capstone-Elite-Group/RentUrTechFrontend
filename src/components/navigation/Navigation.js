import { NavLink, useLocation } from 'react-router-dom';
import Popup from 'reactjs-popup';
import { useRef, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { HiMenuAlt4 } from 'react-icons/hi';
import {
  FaPinterestP, FaFacebookF, FaAngleDoubleLeft,
  FaGooglePlusG, FaVimeoV,
} from 'react-icons/fa';
import { BsTwitter } from 'react-icons/bs';
import { logout } from '../../redux/user/user';
import logo from '../../images/logo1.svg';
import style from './navigation.module.scss';

const Navigation = () => {
  const user = useSelector((state) => state.user);
  const dispatch = useDispatch();
  const menuRef = useRef(null);
  const hamburgerRef = useRef(null);
  const location = useLocation();
  const closedLocation = ['/add-equipment', '/reserve', '/login', '/signup', '/remove-equipment'];

  useEffect(() => {
    if (closedLocation.includes(location.pathname)) {
      menuRef.current.className = style.close_menu;
      hamburgerRef.current.className = style.hamburger_white;
    } else {
      hamburgerRef.current.className = style.hamburger;
    }
  }, [location]);

  const handleMenuClose = () => {
    menuRef.current.className = style.close_menu;
  };
  const handleMenuOpen = () => {
    menuRef.current.className = style.open_menu;
  };
  const handleLogout = () => {
    dispatch(logout());
  };

  return (
    <>
      <div ref={hamburgerRef} className={style.hamburger}>
        <HiMenuAlt4 onClick={handleMenuOpen} />
      </div>
      <aside ref={menuRef}>
        <button type="button" className={style.closing_button} onClick={handleMenuClose}>
          <FaAngleDoubleLeft />
        </button>
        <div className={style.logo}>
          <img src={logo} alt="logo" />
        </div>

        <nav>
          <ul className={style.nav_list}>
            <li>
              <NavLink
                to="/"
                exact
                className={({ isActive }) => (isActive ? style.active : 'inactive')}
              >
                Techs
              </NavLink>
            </li>
            <li>
              <NavLink to="/reserve" className={({ isActive }) => (isActive ? style.active : 'inactive')}>
                Reserve
              </NavLink>
            </li>
            <li>
              <NavLink to="/my-reservations" className={({ isActive }) => (isActive ? style.active : 'inactive')}>
                My Reservations
              </NavLink>
            </li>
            {user?.roles === 'admin' && (
            <>
              <li>
                <NavLink to="/add-equipment" className={({ isActive }) => (isActive ? style.active : 'inactive')}>
                  Add Techs
                </NavLink>
              </li>
              <li>
                <NavLink to="/remove-equipment" className={({ isActive }) => (isActive ? style.active : 'inactive')}>
                  Remove Techs
                </NavLink>
              </li>
            </>
            )}
          </ul>

        </nav>
        <div className={style.social_links}>
          <div className={style.user_login}>
            {user ? (
              <Popup modal trigger={<button type="button" onClick={handleLogout}>Logout</button>}>
                {(close) => (
                  <>
                    <div className={style.overlay}>
                      <div className={style.modal}>
                        <div className={style.logout_message}>
                          Are you sure you want to Logout?
                        </div>
                        <div className={style.logout_choices}>
                          <button type="button" onClick={handleLogout}>Confirm</button>
                          <button
                            type="button"
                            onClick={() => {
                              close();
                            }}
                          >
                            Close
                          </button>

                        </div>
                      </div>
                    </div>
                  </>
                )}
              </Popup>
            ) : (
              <>
                <NavLink to="/login">
                  Login
                </NavLink>
                <NavLink to="/signup">
                  Sign up
                </NavLink>
              </>
            )}
          </div>
        
        </div>
      </aside>
    </>
  );
};

export default Navigation;
