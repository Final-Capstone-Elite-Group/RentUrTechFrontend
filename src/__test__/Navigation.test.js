/* eslint-disable react/display-name */
/* eslint-disable react/prop-types */
import React from 'react';
import { setupServer } from 'msw/node';
import {
  waitFor, render, screen, cleanup,
} from './test-utils';
import Navigation from '../components/navigation/Navigation';
import '@testing-library/jest-dom';
import handlers from './mock/handlers';

const server = setupServer(...handlers);

afterEach(() => cleanup(), server.resetHandlers());

beforeAll(() => server.listen({
  onUnhandledRequest: 'error',
}));

afterAll(() => server.close());

it('Navigation Renders Correctly', async () => {
  const { container } = render(<Navigation />);
  await waitFor(() => screen.getByText('Techs'));
  expect(container).toMatchSnapshot();
});

it('If user is logged out, he can see Login and SignOut buttons', async () => {
  render(<Navigation />);
  expect(screen.getByText('Techs')).toBeInTheDocument();
  expect(screen.getByText('Reservation')).toBeInTheDocument();
  expect(screen.getByText('My Reservations')).toBeInTheDocument();
  expect(screen.getByText('Login')).toBeInTheDocument();
  expect(screen.queryByText('Logout')).not.toBeInTheDocument();
  expect(screen.queryByText('Add Techs')).not.toBeInTheDocument();
});

it('If user is logged in, you can see Techs, Reservation, My Reservations and Logout button', async () => {
  const preloadedState = {
    auth: {
      user: {
        username: 'unique string',
        role: 'user',
      },
      token: '123',
    },
  };
  render(<Navigation />, { preloadedState });
  expect(screen.getByText('Techs')).toBeInTheDocument();
  expect(screen.getByText('Reservation')).toBeInTheDocument();
  expect(screen.getByText('My Reservations')).toBeInTheDocument();
  expect(screen.getByText('Logout')).toBeInTheDocument();
  expect(screen.queryByText('Login')).not.toBeInTheDocument();
  expect(screen.queryByText('Add Techs')).not.toBeInTheDocument();
});

it('If user is admin, he can see Add tech and Remove tech links', async () => {
  const preloadedState = {
    auth: {
      user: {
        username: 'unique string',
        role: 'admin',
      },
      token: '123',
    },
  };
  render(<Navigation />, { preloadedState });
  expect(screen.getByText('Add Techs')).toBeInTheDocument();
  expect(screen.getByText('Remove Techs')).toBeInTheDocument();
});
