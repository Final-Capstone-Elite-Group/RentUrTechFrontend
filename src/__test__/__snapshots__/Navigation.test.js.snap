// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Renders Correctly 1`] = `
<div>
  <div
    class="hamburger"
  >
    <svg
      data-testid="hamburger"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 20 20"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M3 7a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1zM3 13a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
  <aside
    class="relative false"
  >
    <button
      class="closing_button"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M223.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L319.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L393.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34zm-192 34l136 136c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9L127.9 256l96.4-96.4c9.4-9.4 9.4-24.6 0-33.9L201.7 103c-9.4-9.4-24.6-9.4-33.9 0l-136 136c-9.5 9.4-9.5 24.6-.1 34z"
        />
      </svg>
    </button>
    <div
      class="logo"
    >
      <img
        alt="logo"
        src="main_logo.gif"
      />
    </div>
    <nav>
      <ul
        class="nav_list"
      >
        <li>
          <a
            aria-current="page"
            class="active"
            href="/"
          >
            Techs
          </a>
        </li>
        <li>
          <a
            class="inactive"
            href="/reservation"
          >
            Reservation
          </a>
        </li>
        <li>
          <a
            class="inactive"
            href="/my-reservations"
          >
            My Reservations
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="social_links"
    >
      <div
        class="user_login"
      >
        <a
          class=""
          href="/login"
        >
          Login
        </a>
        <a
          class=""
          href="/signup"
        >
          Sign up
        </a>
      </div>
      <ul>
        <li>
          <a
            aria-label="Save"
            href="https://twitter.com/?lang=en"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.026 15c6.038 0 9.341-5.003 9.341-9.334 0-.14 0-.282-.006-.422A6.685 6.685 0 0 0 16 3.542a6.658 6.658 0 0 1-1.889.518 3.301 3.301 0 0 0 1.447-1.817 6.533 6.533 0 0 1-2.087.793A3.286 3.286 0 0 0 7.875 6.03a9.325 9.325 0 0 1-6.767-3.429 3.289 3.289 0 0 0 1.018 4.382A3.323 3.323 0 0 1 .64 6.575v.045a3.288 3.288 0 0 0 2.632 3.218 3.203 3.203 0 0 1-.865.115 3.23 3.23 0 0 1-.614-.057 3.283 3.283 0 0 0 3.067 2.277A6.588 6.588 0 0 1 .78 13.58a6.32 6.32 0 0 1-.78-.045A9.344 9.344 0 0 0 5.026 15z"
              />
            </svg>
          </a>
        </li>
        <li>
          <a
            aria-label="Save"
            href="https://twitter.com/?lang=en"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 320 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M279.14 288l14.22-92.66h-88.91v-60.13c0-25.35 12.42-50.06 52.24-50.06h40.42V6.26S260.43 0 225.36 0c-73.22 0-121.08 44.38-121.08 124.72v70.62H22.89V288h81.39v224h100.17V288z"
              />
            </svg>
          </a>
        </li>
        <li>
          <a
            aria-label="Save"
            href="https://twitter.com/?lang=en"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 640 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M386.061 228.496c1.834 9.692 3.143 19.384 3.143 31.956C389.204 370.205 315.599 448 204.8 448c-106.084 0-192-85.915-192-192s85.916-192 192-192c51.864 0 95.083 18.859 128.611 50.292l-52.126 50.03c-14.145-13.621-39.028-29.599-76.485-29.599-65.484 0-118.92 54.221-118.92 121.277 0 67.056 53.436 121.277 118.92 121.277 75.961 0 104.513-54.745 108.965-82.773H204.8v-66.009h181.261zm185.406 6.437V179.2h-56.001v55.733h-55.733v56.001h55.733v55.733h56.001v-55.733H627.2v-56.001h-55.733z"
              />
            </svg>
          </a>
        </li>
        <li>
          <a
            aria-label="Save"
            href="https://twitter.com/?lang=en"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M447.8 153.6c-2 43.6-32.4 103.3-91.4 179.1-60.9 79.2-112.4 118.8-154.6 118.8-26.1 0-48.2-24.1-66.3-72.3C100.3 250 85.3 174.3 56.2 174.3c-3.4 0-15.1 7.1-35.2 21.1L0 168.2c51.6-45.3 100.9-95.7 131.8-98.5 34.9-3.4 56.3 20.5 64.4 71.5 28.7 181.5 41.4 208.9 93.6 126.7 18.7-29.6 28.8-52.1 30.2-67.6 4.8-45.9-35.8-42.8-63.3-31 22-72.1 64.1-107.1 126.2-105.1 45.8 1.2 67.5 31.1 64.9 89.4z"
              />
            </svg>
          </a>
        </li>
        <li>
          <a
            aria-label="Save"
            href="https://twitter.com/?lang=en"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 384 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M204 6.5C101.4 6.5 0 74.9 0 185.6 0 256 39.6 296 63.6 296c9.9 0 15.6-27.6 15.6-35.4 0-9.3-23.7-29.1-23.7-67.8 0-80.4 61.2-137.4 140.4-137.4 68.1 0 118.5 38.7 118.5 109.8 0 53.1-21.3 152.7-90.3 152.7-24.9 0-46.2-18-46.2-43.8 0-37.8 26.4-74.4 26.4-113.4 0-66.2-93.9-54.2-93.9 25.8 0 16.8 2.1 35.4 9.6 50.7-13.8 59.4-42 147.9-42 209.1 0 18.9 2.7 37.5 4.5 56.4 3.4 3.8 1.7 3.4 6.9 1.5 50.4-69 48.6-82.5 71.4-172.8 12.3 23.4 44.1 36 69.3 36 106.2 0 153.9-103.5 153.9-196.8C384 71.3 298.2 6.5 204 6.5z"
              />
            </svg>
          </a>
        </li>
      </ul>
      <p
        class="copy"
      >
         Â© Copyright 2022, RUT incorporated
      </p>
    </div>
  </aside>
</div>
`;
